src/VecSim/algorithms/hnsw/hnsw_factory.cpp

```diff
@@ -119,4 +119,84 @@ size_t EstimateElementSize(const HNSWParams *params) {
      */
     return size_meta_data + size_total_data_per_element;
 }
+
+#ifdef BUILD_TESTS
+
+template <typename DataType, typename DistType = DataType>
+inline VecSimIndex *NewIndex_ChooseMultiOrSingle(std::ifstream &input, const HNSWParams *params,
+                                                 std::shared_ptr<VecSimAllocator> allocator,
+                                                 Serializer::EncodingVersion version) {
+    VecSimIndex *index = nullptr;
+    // check if single and call the ctor that loads index information from file.
+    if (params->multi)
+        index =
+            new (allocator) HNSWIndex_Multi<DataType, DistType>(input, params, allocator, version);
+    else
+        index =
+            new (allocator) HNSWIndex_Single<DataType, DistType>(input, params, allocator, version);
+
+    reinterpret_cast<HNSWIndex<DataType, DistType> *>(index)->restoreGraph(input);
+
+    return index;
+}
+
+void InitializeParams(std::ifstream &input, HNSWParams &params) {
+    Serializer::readBinaryPOD(input, params.dim);
+    Serializer::readBinaryPOD(input, params.type);
+    Serializer::readBinaryPOD(input, params.metric);
+    Serializer::readBinaryPOD(input, params.blockSize);
+    Serializer::readBinaryPOD(input, params.multi);
+    Serializer::readBinaryPOD(input, params.epsilon);
+}
+
+VecSimIndex *NewIndex(const std::string &location, const HNSWParams *params) {
+
+    std::ifstream input(location, std::ios::binary);
+    if (!input.is_open()) {
+        throw std::runtime_error("Cannot open file");
+    }
+
+    Serializer::EncodingVersion version = Serializer::ReadVersion(input);
+    HNSWParams file_params;
+    switch (version) {
+    case Serializer::EncodingVersion_V2: {
+        // Algorithm type is only serialized from V2 up.
+        VecSimAlgo algo = VecSimAlgo_INVALID;
+        Serializer::readBinaryPOD(input, algo);
+        if (algo != VecSimAlgo_HNSWLIB) {
+            input.close();
+            throw std::runtime_error("Cannot load index: bad algorithm type");
+        }
+        // this information is serialized from V2 and up
+        InitializeParams(input, file_params);
+        break;
+    }
+    case Serializer::EncodingVersion_V1: {
+        assert(params);
+        file_params.type = params->type;
+        file_params.dim = params->dim;
+        file_params.metric = params->metric;
+        file_params.multi = params->multi;
+        file_params.blockSize = params->blockSize;
+        file_params.epsilon = params->epsilon;
+        break;
```

Comment:
Good question
maybe epsilon and block size can have default values because they don't affect the loading process, but all the others must be the same as the loaded index since we cant get them from the file.